### PR TITLE
[stable/traefik] Allow annotations and labels on Service and UI Ingress

### LINKED ISSUE #334: Allow traefik Service to be configured with the annotations necessary for AWS ELB
As documented here - https://github.com/kubernetes/kubernetes/blob/master/docs/design/aws_under_the_hood.md#nodeport-and-loadbalancer-services - if you want AWS ELB configured you need to add certain annotations to the service.

The `Service` in the Traefik chart should allow setting these via `values.yaml` overrides.

Thank you!

### PR DESCRIPTION
**Pull request by cknowles**
Should resolve https://github.com/kubernetes/charts/issues/334.

Allows for many additional use cases such as setup of various AWS ELB
and Route 53 annotations including use of
https://github.com/fairfaxmedia/area53.

I've avoided bumping the Chart version for now to avoid clashes with other PRs.

cc @krancour the chart maintainer

### CODE REVIEW

File: stable/traefik/README.md [hunk 1]
```diff
@@ -87,6 +87,9 @@ The following tables lists the configurable parameters of the Traefik chart and
 | `acme.persistence.size`         | Minimum size of the volume requested                                 | `1Gi`                                     |
 | `dashboard.enabled`             | Whether to enable the Traefik dashboard                              | `false`                                   |
 | `dashboard.domain`              | Domain for the Traefik dashboard                                     | `traefik.example.com`                     |
+| `dashboard.ingress.annotations` | Annotations for the Traefik dashboard Ingress definition             | None                                      |
+| `service.annotations`           | Annotations for the Traefik Service definition                       | None                                      |
+| `service.labels`                | Additional labels for the Traefik Service definition                 | None                                      |
```
Comment (krancour) [inline stable/traefik/README.md]: Can we make it clear that these are maps?

**Review by krancour — CHANGES_REQUESTED** [review]

**Comment by krancour** [comment]
This looks pretty good. I do want to manually test this.

File: stable/traefik/templates/dashboard-ingress.yaml [hunk 2]
```diff
@@ -8,6 +8,10 @@ metadata:
     chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
     release: "{{ .Release.Name }}"
     heritage: "{{ .Release.Service }}"
+  annotations:
+  {{- range $key, $value := .Values.dashboard.ingress.annotations }}
+    {{ $key }}: {{ $value | quote }}
+  {{- end }}
```
Comment (krancour) [inline stable/traefik/templates/dashboard-ingress.yaml]: Testing this, I found that you're really going to have to wrap this in a double-if condition similar to what you have done in #414.

File: stable/traefik/templates/service.yaml [hunk 3]
```diff
@@ -7,6 +7,13 @@ metadata:
     chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
     release: "{{ .Release.Name }}"
     heritage: "{{ .Release.Service }}"
+  {{- range $key, $value := .Values.service.labels }}
+    {{ $key }}: {{ $value | quote }}
+  {{- end }}
```
Comment (krancour) [inline stable/traefik/templates/service.yaml]: Testing this, I found that you're really going to have to wrap this in a double-if condition similar to what you have done in #414.

File: stable/traefik/templates/service.yaml [hunk 4]
```diff
@@ -7,6 +7,13 @@ metadata:
...
+  annotations:
+  {{- range $key, $value := .Values.service.annotations }}
+    {{ $key }}: {{ $value | quote }}
+  {{- end }}
```
Comment (krancour) [inline stable/traefik/templates/service.yaml]: Testing this, I found that you're really going to have to wrap this in a double-if condition similar to what you have done in #414.

**Review by krancour — CHANGES_REQUESTED** [review]

**Comment by cknowles** [comment]
@krancour I've updated those, there's still an assumption that `dashboard` exists but I think that's a fair assumption. I hope golang templates improve in that regard or Helm adds some custom functions to deal with this.

**Comment by krancour** [comment]
> there's still an assumption that `dashboard` exists

That's an easier assumption to make, because a `dashboard` section has been present in the default `values.yaml` since the very first version of this chart-- which means it necessarily exists in anyone's existing (custom) `values.yaml`. By contrast, anyone who uses their existing `values.yaml` with the new / improved chart would be missing the sections we added the extra guards for. It's those folks that I wanted to look out for here. Backwards compatibility between the updated chart and old `values.yaml` is the name of the game.

**Comment by krancour** [comment]
@c-knowles, I've manually tested this and it works great.

One question... do you think there's any value in allowing labels and annotations to be set the same way on the dashboard's service as well (if the dashboard is enabled)?

**Comment by cknowles** [comment]
@krancour could do if you wanted. I can't think of many use cases but perhaps some companies have requirements over labels, for example we like to label apps with a tier.

**Comment by krancour** [comment]
@c-knowles up to you. If you can rationalize a use case for, add it. If not, then don't bother.

This LGTM either way.

**Comment by cknowles** [comment]
@krancour most of the ones I've thought of so far overlap with https://github.com/kubernetes/helm/issues/1800. I'd much prefer for label uniformity issues to be dealt with at a higher level rather than every chart. As for annotations, I can't spot much value right now. Let's skip it until a use case comes up?

**Comment by krancour** [comment]
SGTM @c-knowles

**Comment by viglesiasce** [comment]
@k8s-bot ok to test

**Pull request merged**